amethyst_error/src/lib.rs

```diff
@@ -0,0 +1,402 @@
+//! Contains the `Error` type and company as used by Amethyst.
+//!
+//! **Note:** This type is not intended to be used outside of Amethyst.
+//! If you are integrating a crate from amethyst to use this, it is recommended that you treat this
+//! type as an opaque [`std::error::Error`].
+//!
+//! [`std::error::Error`]: https://doc.rust-lang.org/std/error/trait.Error.html
+use self::internal::{Backtrace, HAS_BACKTRACE};
+use std::borrow::Cow;
+use std::env;
+use std::error;
+use std::ffi;
+use std::fmt;
+use std::result;
+use std::sync::atomic;
+
+/// Private
+enum ErrorImpl {
+    Message(Cow<'static, str>),
+    Error(Box<dyn error::Error + Send + Sync>),
+}
+
+impl fmt::Debug for ErrorImpl {
+    fn fmt(&self, fmt: &mut fmt::Formatter) -> fmt::Result {
+        match *self {
+            ErrorImpl::Message(ref message) => fmt::Debug::fmt(message, fmt),
+            ErrorImpl::Error(ref e) => fmt::Debug::fmt(e, fmt),
+        }
+    }
+}
+
+impl From<&'static str> for ErrorImpl {
+    fn from(message: &'static str) -> Self {
+        ErrorImpl::Message(Cow::from(message))
+    }
+}
+
+impl From<String> for ErrorImpl {
+    fn from(message: String) -> Self {
+        ErrorImpl::Message(Cow::from(message))
+    }
+}
+
+impl From<Box<dyn error::Error + Send + Sync>> for ErrorImpl {
+    fn from(error: Box<dyn error::Error + Send + Sync>) -> Self {
+        ErrorImpl::Error(error)
+    }
+}
+
+/// The error type used by Amethyst.
+///
+/// Wraps error diagnostics like messages and other errors, and keeps track of causal chains and
+/// backtraces.
+pub struct Error {
+    error_impl: ErrorImpl,
```

Comment:
Hmm, this means the error is bigger than 3 times pointer size.

I think the best strategy is, independent of how std will adopt failure, to have two kinds of errors:

* cheap to construct enums for things that may fail but can easily be recovered from
* allocated, pointer sized errors for things that usually don't fail as to not slow down the succesful execution

I'll explain this more later today.

Reply:
To add to what torkleyy says.

It might be better to have common `Error` as

```rust
pub enum Error {
    Thin(PlainError),
    Fat(FatError)
}
```

Where `FatError` could contains extra information while `PlainError` would be cheap enum (most likely used by `amethyst` APIs)

Reply:
The cheap errors are better built locally in the crate that uses them, instead of having one centralized type.

Reply:
Yes, definitely.

Reply:
That is, if by "cheap" you mean enums. Both variants are cheap in their own way.

Reply:
Ok I thought this crate would introduce general error for all `amethyst` crates, if cheap errors will remain in other crates then I don't see this bloated error as bad thing

Reply:
But this `Error` is even fatter than `FatError`
```
pub enum Error {
    Thin(PlainError),
    Fat(FatError)
}
```

Reply:
It replaces heap allocations at cost of single integer

Reply:
@udoprog To explain my suggestion more, I'd imagine that `Erorr` is only a `Box` around `ErrorImpl<E>` with `E: Error + ?Sized`. At least that's the pattern I prefer for this kind of things, because that way all the error data is on the heap, while at the same time there's only one allocation (per fat error).

Reply:
Oh maybe @torkleyy approach would be better, if we assume that errors are rare anyway

Reply:
@torkleyy I'm gonna try to wrap my head around what you are proposing. I'll poke you on Discord.